src/main/java/ua/tc/marketplace/service/impl/AdServiceImpl.java

```diff
@@ -89,4 +110,63 @@ public void deleteAd(Long adId) {
   private Ad getAd(Long adId) {
     return adRepository.findById(adId).orElseThrow(() -> new AdNotFoundException(adId));
   }
+
+  private AdAttribute getAdAttribute(Entry<Long, String> entry, Ad finalAd, Category category) {
+    return new AdAttribute(
+        null,
+        finalAd,
+        category.getAttributes().stream()
+            .filter(attribute -> attribute.getId().equals(entry.getKey()))
+            .findFirst()
+            .orElseThrow(() -> new AttributeNotFoundException(entry.getKey())),
+        entry.getValue());
+  }
+
+  private List<AdAttributeRequestDto> parseJsonAdAttributes(CreateAdDto dto) {
+    ObjectMapper mapper = new ObjectMapper();
+    try {
+      return mapper.readValue(dto.adAttributes(), new TypeReference<>() {});
+    } catch (IOException e) {
+      throw new FailedToParseAdAttributesJsonException(dto.adAttributes());
+    }
+  }
+
+  private List<AdAttribute> mapToAdAttributes(
+      List<AdAttributeRequestDto> adAttributeRequestDtos, Category category, Ad ad) {
+
+    Map<Long, String> attributeMap =
+        adAttributeRequestDtos.stream()
+            .collect(
+                Collectors.toMap(AdAttributeRequestDto::attributeId, AdAttributeRequestDto::value));
+    // Convert the list of attributes to a set of IDs
+    Set<Long> requiredAttributeIds =
+        category.getAttributes().stream().map(Attribute::getId).collect(Collectors.toSet());
```

Comment:
Note to think about: maybe it makes sense to add a flag `isMandatory` to `Attribute`. And filter for mandatory attributes only here. Because currently we have all attributes mandatory by default, which might be not convenient in some cases

Reply:
We are based on thought that Category contains a List of Required attributes. So when we create Ad we check if AdAttributes of this Ad match a list of required attributes from this category